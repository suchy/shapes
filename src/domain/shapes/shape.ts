export enum ShapeColors {
  purple = 'purple',
  red = 'red',
  green = 'green',
  blue = 'blue',
  yellow = 'yellow',
  grey = 'grey'
}

export enum ShapeShapes {
  round = 'round',
  square = 'square',
  triangle = 'triangle',
  oval = 'oval',
  rectangle = 'rectangle'
}

export interface Shape {
  id: number;
  color: ShapeColors;
  shape: ShapeShapes;
}

interface ShapeFactoryInput {
  id: number;
  color: ShapeColors;
  shape: ShapeShapes;
}

export const shapeFactory = ({ id, color, shape }: ShapeFactoryInput) => {
  if (!id) {
    throw new Error('Shape id is invalid');
  }

  const colors = Object.keys(ShapeColors);

  if (!color || !colors.includes(color)) {
    throw new Error('Shape color is invalid');
  }

  const shapes = Object.keys(ShapeShapes);

  if (!shape || !shapes.includes(shape)) {
    throw new Error('Shape shape is invalid');
  }

  const createdShape: Shape = { id, color, shape };

  return createdShape;
};
