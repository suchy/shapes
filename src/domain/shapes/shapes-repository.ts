import { apiClient } from '../../infrastructure/api-client';
import { shapeFactory, Shape } from './shape';

export const shapesRepository = {
  async getShapes(): Promise<Shape[]> {
    try {
      const response = await apiClient.get('/shapes.json');

      const rawShapes = response.data;

      // eslint-disable-next-line
      const shapes = rawShapes.map((rawShape: any) =>
        shapeFactory({
          id: rawShape.id,
          color: rawShape.color,
          shape: rawShape.shape
        })
      );

      return shapes;
    } catch (error) {
      // eslint-disable-next-line
      console.error(error);
      throw error;
    }
  }
};
