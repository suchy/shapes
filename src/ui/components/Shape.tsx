import styled from 'styled-components';

import { Shape as ShapeType } from '../../domain/shapes/shape';

interface ShapeProps {
  className?: string;
  shape: ShapeType;
}

const ShapeBase = styled.div<{ color: ShapeType['color'] }>`
  background: ${({ color }) => color};
  overflow: hidden;
`;

const Round = styled(ShapeBase)`
  border-radius: 50%;
`;

const Square = styled(ShapeBase)``;

const Triangle = styled(ShapeBase)`
  background: transparent;

  & > div {
    position: relative;
    top: 25%;
    height: 50%;
    overflow: hidden;

    &::after {
      content: '';
      display: block;
      width: 100%;
      height: 180%;
      transform: rotate(-45deg);
      position: absolute;
      left: -3%;
      top: 45%;
      background: ${({ color }) => color};
    }
  }
`;

const Oval = styled(ShapeBase)`
  background: transparent;

  & > div {
    width: 100%;
    height: 80%;
    border-radius: 50%;
    position: relative;
    top: 10%;
    background: ${({ color }) => color};
  }
`;

const Rectangle = styled(ShapeBase)`
  background: transparent;

  & > div {
    width: 100%;
    height: 70%;
    position: relative;
    top: 15%;
    background: ${({ color }) => color};
  }
`;

export const Shape = ({ className, shape }: ShapeProps) => {
  const componentsMap = {
    round: Round,
    square: Square,
    triangle: Triangle,
    oval: Oval,
    rectangle: Rectangle
  };

  const ShapeComponent = componentsMap[shape.shape];

  return (
    <ShapeComponent className={className} color={shape.color}>
      <div />
    </ShapeComponent>
  );
};
