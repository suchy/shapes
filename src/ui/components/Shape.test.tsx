import { render } from '@testing-library/react';

import { ShapeColors, ShapeShapes } from '../../domain/shapes/shape';

import { Shape } from './Shape';

describe('Shape', () => {
  it('should match snapshot', () => {
    const shapes = Object.keys(ShapeShapes);
    const colors = Object.keys(ShapeColors);

    shapes.forEach((shape) => {
      colors.forEach((color) => {
        const shapeData = { id: 1, color, shape };
        // eslint-disable-next-line
        // @ts-ignore
        const { container } = render(<Shape shape={shapeData} />);
        expect(container.firstChild).toMatchSnapshot();
      });
    });
  });
});
