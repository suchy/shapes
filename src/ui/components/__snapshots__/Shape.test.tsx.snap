// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shape should match snapshot 1`] = `
<div
  class="sc-bdfBwQ sc-gsTCUz bEPRz kmBfpU"
  color="purple"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 2`] = `
<div
  class="sc-bdfBwQ sc-gsTCUz fLKWre kmBfpU"
  color="red"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 3`] = `
<div
  class="sc-bdfBwQ sc-gsTCUz gMtiFK kmBfpU"
  color="green"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 4`] = `
<div
  class="sc-bdfBwQ sc-gsTCUz cIXVGL kmBfpU"
  color="blue"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 5`] = `
<div
  class="sc-bdfBwQ sc-gsTCUz kxyUnR kmBfpU"
  color="yellow"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 6`] = `
<div
  class="sc-bdfBwQ sc-gsTCUz fLCdOs kmBfpU"
  color="grey"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 7`] = `
<div
  class="sc-bdfBwQ sc-dlfnbm bEPRz"
  color="purple"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 8`] = `
<div
  class="sc-bdfBwQ sc-dlfnbm fLKWre"
  color="red"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 9`] = `
<div
  class="sc-bdfBwQ sc-dlfnbm gMtiFK"
  color="green"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 10`] = `
<div
  class="sc-bdfBwQ sc-dlfnbm cIXVGL"
  color="blue"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 11`] = `
<div
  class="sc-bdfBwQ sc-dlfnbm kxyUnR"
  color="yellow"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 12`] = `
<div
  class="sc-bdfBwQ sc-dlfnbm fLCdOs"
  color="grey"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 13`] = `
<div
  class="sc-bdfBwQ sc-hKgILt bEPRz bfjoYk"
  color="purple"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 14`] = `
<div
  class="sc-bdfBwQ sc-hKgILt fLKWre dIJjBd"
  color="red"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 15`] = `
<div
  class="sc-bdfBwQ sc-hKgILt gMtiFK cCWxIV"
  color="green"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 16`] = `
<div
  class="sc-bdfBwQ sc-hKgILt cIXVGL cbeKOc"
  color="blue"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 17`] = `
<div
  class="sc-bdfBwQ sc-hKgILt kxyUnR biKfYG"
  color="yellow"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 18`] = `
<div
  class="sc-bdfBwQ sc-hKgILt fLCdOs fRpkxj"
  color="grey"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 19`] = `
<div
  class="sc-bdfBwQ sc-eCssSg bEPRz eRvqwA"
  color="purple"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 20`] = `
<div
  class="sc-bdfBwQ sc-eCssSg fLKWre hfvboh"
  color="red"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 21`] = `
<div
  class="sc-bdfBwQ sc-eCssSg gMtiFK koROfJ"
  color="green"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 22`] = `
<div
  class="sc-bdfBwQ sc-eCssSg cIXVGL gWUKLE"
  color="blue"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 23`] = `
<div
  class="sc-bdfBwQ sc-eCssSg kxyUnR isoRIa"
  color="yellow"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 24`] = `
<div
  class="sc-bdfBwQ sc-eCssSg fLCdOs csZguL"
  color="grey"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 25`] = `
<div
  class="sc-bdfBwQ sc-jSgupP bEPRz fyeLUR"
  color="purple"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 26`] = `
<div
  class="sc-bdfBwQ sc-jSgupP fLKWre bLUvMg"
  color="red"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 27`] = `
<div
  class="sc-bdfBwQ sc-jSgupP gMtiFK fmSRmM"
  color="green"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 28`] = `
<div
  class="sc-bdfBwQ sc-jSgupP cIXVGL bQutAR"
  color="blue"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 29`] = `
<div
  class="sc-bdfBwQ sc-jSgupP kxyUnR cmlrvf"
  color="yellow"
>
  <div />
</div>
`;

exports[`Shape should match snapshot 30`] = `
<div
  class="sc-bdfBwQ sc-jSgupP fLCdOs kJVjSu"
  color="grey"
>
  <div />
</div>
`;
